Objects/frameobject.c

```diff
@@ -1430,4 +1430,33 @@ _PyEval_BuiltinsFromGlobals(PyThreadState *tstate, PyObject *globals)
     return _PyEval_GetBuiltins(tstate);
 }
 
+PyObject *
+PyFrame_GetVar(PyFrameObject *frame, PyObject *name)
+{
+    PyObject *locals = PyFrame_GetLocals(frame);
+    if (locals == NULL) {
+        return NULL;
+    }
+    PyObject *value = PyDict_GetItemWithError(locals, name);
+    Py_DECREF(locals);
+
+    if (value == NULL) {
+        if (PyErr_ExceptionMatches(PyExc_KeyError)) {
+            PyErr_Clear();
```

Comment:
I'm sorry, I don't like this at all. Didn't we even have a discussion at some point about API design anti-patterns where returning NULL while sometimes setting an exception and sometimes not was generally agreed to be something to avoid?

In any case, the caller still has to check the exception, so we're not saving them any work, and if they want to be able to distinguish between "variable not set" and other errors they can check for a specific exception.

I'd say the appropriate exception to set here is probably `NameError`.

Also note that there seems to be a bug in this code? `PyDict_GetItemWithError()` doesn't set `KeyError` when the key is not present -- it has the behavior I am unhappy about here. So why are you checking for `KeyError`?

Reply:
Oh right this is agood point.

I also don't think it's good that there's a mix of `NULL` without error and `NULL` with error. This is really confusing design.